In [0]:
import pandas as pd

In [0]:
from google.colab import files
uploaded = files.upload()

Saving weatherHistory.csv to weatherHistory.csv


In [0]:
weather_hist = pd.read_csv('weatherHistory.csv')

In [0]:
weather_hist.head(5)

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

Найти самый холодный и самый ветреный дни за всю историю



Нас интересует день с минимальной температурой - допустим, измеренной, а не ощущаемой (-21.822222) и день с максимальной скоростью ветра (63.852600)




In [0]:
weather_hist.loc[weather_hist['Temperature (C)'] == -21.822222]

Empty DataFrame
Columns: [Formatted Date, Summary, Precip Type, Temperature (C), Apparent Temperature (C), Humidity, Wind Speed (km/h), Wind Bearing (degrees), Visibility (km), Loud Cover, Pressure (millibars), Daily Summary]
Index: []

Ничего хорошего не получилось; возможно, значение в таблице было округленным, а на самом деле знаков после запятой больше

In [0]:
weather_hist[weather_hist['Temperature (C)'] < -21]

Formatted Date  ...                                      Daily Summary
54847  2012-02-10 07:00:00.000 +0100  ...  Foggy starting in the morning continuing until...
55490  2012-02-09 02:00:00.000 +0100  ...                               Foggy until morning.
55493  2012-02-09 05:00:00.000 +0100  ...                               Foggy until morning.
55494  2012-02-09 06:00:00.000 +0100  ...                               Foggy until morning.

[4 rows x 12 columns]

Самый холодный день - 10 февраля 2012
И уже интересно, как считали температуру, чтобы в другие 3 чуть менее холодных дня получить в дробной части 1 в периоде и ровно столько же в колонке "ощущается как" - при том, что прочие показатели не совпадают. Единственное, что приходит в голову - возможно, переводили из какой-то другой шкалы (Фаренгейт?), где числа были с меньшим количеством знаков после запятой

Самый ветреный день - 29 января 2007.

In [0]:
weather_hist[weather_hist['Wind Speed (km/h)'] > 63]

Formatted Date  ...                                      Daily Summary
12181  2007-01-29 13:00:00.000 +0100  ...  Mostly cloudy throughout the day and windy sta...

[1 rows x 12 columns]

сколько раз шел дождь летом 2014 года

In [0]:
weather_hist['datetime'] = pd.to_datetime(weather_hist['Formatted Date'], utc=True)

In [0]:
weather_hist.dtypes

Formatted Date                           object
Summary                                  object
Precip Type                              object
Temperature (C)                         float64
Apparent Temperature (C)                float64
Humidity                                float64
Wind Speed (km/h)                       float64
Wind Bearing (degrees)                  float64
Visibility (km)                         float64
Loud Cover                              float64
Pressure (millibars)                    float64
Daily Summary                            object
datetime                    datetime64[ns, UTC]
dtype: object

In [0]:
weather_hist_2014 = weather_hist.loc[weather_hist['datetime'].dt.year == 2014]

In [0]:
weather_hist_summer2014 = weather_hist2014.loc[weather_hist2014['datetime'].dt.month > 5]
weather_hist_summer2014 = weather_hist_summer2014.loc[weather_hist_summer2014['datetime'].dt.month < 9]

In [0]:
weather_hist_summer2014_rain = weather_hist_summer2014.loc[weather_hist_summer2014['Precip Type'] == 'rain']

In [0]:
weather_hist_summer2014_rain.shape

(2205, 13)

Получается какая-то ерунда, не было летом 2014 года 2205 дней, даже если дождь шел беспрерывно

---



In [0]:
weather_hist_summer2014_rain

Formatted Date  ...                  datetime
70872  2014-08-01 00:00:00.000 +0200  ... 2014-07-31 22:00:00+00:00
70873  2014-08-01 01:00:00.000 +0200  ... 2014-07-31 23:00:00+00:00
70874  2014-08-01 02:00:00.000 +0200  ... 2014-08-01 00:00:00+00:00
70875  2014-08-01 03:00:00.000 +0200  ... 2014-08-01 01:00:00+00:00
70876  2014-08-01 04:00:00.000 +0200  ... 2014-08-01 02:00:00+00:00
...                              ...  ...                       ...
75234  2014-06-09 21:00:00.000 +0200  ... 2014-06-09 19:00:00+00:00
75235  2014-06-09 22:00:00.000 +0200  ... 2014-06-09 20:00:00+00:00
75236  2014-06-09 23:00:00.000 +0200  ... 2014-06-09 21:00:00+00:00
78189  2014-09-01 00:00:00.000 +0200  ... 2014-08-31 22:00:00+00:00
78190  2014-09-01 01:00:00.000 +0200  ... 2014-08-31 23:00:00+00:00

[2205 rows x 13 columns]

Оказывается, каждый день проводилось несколько наблюдений. Причём посчитать уникальные дни недостаточно: к сожалению, каждый месяц нумерация начинается заново. Поэтому добавим колонку, которая будет одинаковой для измерений, сделанных в один день

In [0]:
weather_hist_summer2014_rain['date'] = weather_hist_summer2014_rain['datetime'].dt.date

In [0]:
len(weather_hist_summer2014_rain['date'].unique())

92

Летом 2014 года дождь шёл 92 дня, то есть каждый день.
Странно. Искать измерения, когда дождь не шёл, чтобы проверить этот результат, бесполезно, так как в один и тот же день дождь мог утром идти, а вечером нет. Если мы поделим длину изначальной таблицу (2205) на 24 часа в сутках (кажется, наблюдения делались раз в час), получится как раз 92. Видимо, дождь действительно шёл беспрерывно. Наверно, при таком результате надо приходить к выводу, что колонку "осадки" я перезаписала вместо того, чтобы выбрать дождливые дни, но, кажется, нет. Какое замечательное лето.

Найдём самый ветреный месяц




In [0]:
weather_hist['month'] = weather_hist['datetime'].dt.month

In [0]:
weather_hist.groupby('month').agg(['mean'])

Temperature (C) Apparent Temperature (C)  ... Loud Cover Pressure (millibars)
                 mean                     mean  ...       mean                 mean
month                                           ...                                
1            0.815678                -1.938761  ...        0.0          1006.245222
2            2.166880                -0.555908  ...        0.0          1003.925126
3            6.914976                 5.082583  ...        0.0          1001.564999
4           12.769200                12.158287  ...        0.0          1009.995154
5           16.874455                16.733265  ...        0.0          1003.376970
6           20.728510                20.755415  ...        0.0          1002.019582
7           22.965462                22.977605  ...        0.0          1008.562854
8           22.339134                22.264600  ...        0.0          1001.470591
9           17.503458                17.389856  ...        0.0          1000.835433
10          11.334119                10.684673  ...        0.0          1003.099488
11           6.587549                 5.020255  ...        0.0          1012.420129
12           1.625086                -0.960443  ...        0.0           985.781311

[12 rows x 8 columns]

Средняя скорость ветра выше всего в марте (13.4)

In [0]:
weather_hist.drop(['month'], axis=1, inplace=True)

Найдём самый влажный год

In [0]:
weather_hist['year'] = weather_hist['datetime'].dt.year

Определим, какие года представлены в таблице

In [0]:
weather_hist['year'].unique()

array([2006, 2005, 2010, 2007, 2008, 2009, 2011, 2012, 2013, 2014, 2015,
       2016])

C 2005 по 2016, без пропусков
Посмотрим, какие варианты заполнения колонки "осадки" встречаются

In [0]:
weather_hist['Precip Type'].unique()

array(['rain', 'snow', nan], dtype=object)

Видимо, влажности соответствует что угодно, кроме nan

In [0]:
most_humid_year = 0
max_humid_hours = 0
for y in range(2005, 2017):
  weather_year = weather_hist.loc[weather_hist['datetime'].dt.year == y]
  weather_year_rain = weather_year.loc[weather_year['Precip Type'] == 'rain']
  weather_year_snow = weather_year.loc[weather_year['Precip Type'] == 'snow']
  humid_hours = len(weather_year_rain['datetime'].unique()) + len(weather_year_snow['datetime'].unique()) 
  if humid_hours > max_humid_hours:
    max_humid_hours = humid_hours
    most_humid_year = y
print(y)

2016


Самый влажный год - 2016


In [0]:
weather_hist.drop(['year'], axis=1, inplace=True)

Найдём неделю с самой переменчивой погодой

In [0]:
weather_hist['week'] = weather_hist['datetime'].dt.week

Если ориентироваться на температуру, самая переменчивая погода бывает на 34 неделе года

In [0]:
weather_hist.groupby('week').agg(['std'])

Temperature (C) Apparent Temperature (C)  ... Loud Cover Pressure (millibars)
                 std                      std  ...        std                  std
week                                           ...                                
1           5.008042                 5.530550  ...        0.0           246.685649
2           4.811115                 5.394467  ...        0.0            89.103028
3           4.532339                 5.165160  ...        0.0            67.561625
4           5.482606                 6.102784  ...        0.0            42.686836
5           5.251543                 5.925691  ...        0.0           122.760869
6           6.456311                 7.246163  ...        0.0           110.584848
7           4.628565                 5.535001  ...        0.0           160.344471
8           4.967552                 5.908516  ...        0.0            48.021065
9           4.839419                 5.658514  ...        0.0           105.512072
10          4.947341                 6.001244  ...        0.0           113.255560
11          5.265073                 6.399644  ...        0.0           122.570227
12          5.452593                 6.789302  ...        0.0           146.403166
13          5.334651                 6.508104  ...        0.0           107.712620
14          5.272616                 6.310754  ...        0.0            41.245168
15          5.270850                 6.040311  ...        0.0            94.238453
16          5.020456                 5.607845  ...        0.0             6.580682
17          5.413772                 5.765406  ...        0.0            78.377524
18          5.522132                 5.814538  ...        0.0           166.223026
19          5.025860                 5.253456  ...        0.0            81.778383
20          5.262296                 5.506798  ...        0.0            24.043112
21          4.923256                 4.905467  ...        0.0            74.507895
22          5.221413                 5.287079  ...        0.0            99.612098
23          5.599345                 5.678155  ...        0.0            33.686783
24          5.158063                 5.128353  ...        0.0           130.318992
25          5.794001                 6.013124  ...        0.0           166.398242
26          5.538135                 5.616053  ...        0.0            74.544189
27          5.677395                 5.604498  ...        0.0            52.981288
28          5.755464                 5.898430  ...        0.0           105.067709
29          5.864455                 5.895502  ...        0.0           102.531259
30          5.587042                 5.563011  ...        0.0            23.811323
31          5.383029                 5.370217  ...        0.0            74.070267
32          5.709605                 5.716168  ...        0.0           107.605653
33          5.925675                 5.762977  ...        0.0            57.886752
34          6.149509                 5.916474  ...        0.0           121.962378
35          5.875341                 5.673549  ...        0.0           174.319196
36          6.043471                 5.967835  ...        0.0            99.924434
37          5.163760                 5.136622  ...        0.0           174.367014
38          5.689703                 5.848190  ...        0.0           153.414098
39          5.315730                 5.388077  ...        0.0            24.364101
40          5.486404                 5.932160  ...        0.0           122.502358
41          5.161809                 5.596601  ...        0.0           108.252108
42          5.234595                 6.054321  ...        0.0            78.628116
43          4.911441                 5.653080  ...        0.0           118.073410
44          5.231834                 6.106931  ...        0.0           181.023391
45          4.951843                 5.719654  ...        0.0            53.756782
46          4.762689 

## Регрессия на тексте

In [2]:
from google.colab import files
uploaded = files.upload()

Saving data.csv to data.csv


In [0]:
import pandas as pd

In [0]:
data = pd.read_csv('data.csv')

In [0]:
import numpy as np
from sklearn.metrics import mean_squared_error

In [0]:
data['text_len'] = data['description'].apply(len)
data['text_len_tokens'] = data['description'].apply(lambda x: len(x.split()))
data['number_of_caps'] = data['description'].apply(lambda x: len([ch for ch in x if ch.isupper()]))
data['number_of_nonalphanum'] = data['description'].apply(lambda x: len([ch for ch in x if not ch.isalnum()]))

### Домашнее задание 2

Попробуйте придумать признаки (как минимум 5), которые можно извлечь из текста, помимо тех 4, которые мы уже использовали. Обучите модель на них и проверьте, улучашается ли качество. Опишите признаки, которое придумали и почему считаете, что они должны показывать вероятность покупки.

Побробуйте подобрать параметры векторизатора (TfidfVectorizer или CountVectorizer, или обоих) и параметры алгоритма так, чтобы качество, которое у нас получилось (0.248) улучшилось. Опишите то, что попробовали.


Признак 1: количество чисел и последовательностей букв, обозначающих размеры одежды. Характеризует точность описания

In [0]:
import re
def size_in_text(text):
  n = len(re.findall('([0-9]+)([\.,]?)([0-9]+)?', text)) + len(re.findall('X*[SML]', text))
  return(n)


In [0]:
data['size_info'] = data['description'].apply(lambda x: size_in_text(x))

Признак 2: количесиво вопросительных и восклицательных знаков. Предполагается, что, чем их больше, тем более ненейтрально описаник

In [0]:
def question_and_exclamation_marks(text):
  question = text.count('?')
  exclamation = text.count('!')
  n = question + exclamation
  return(n)

In [0]:
data['question_and_exclamation'] = data['description'].apply(lambda x: question_and_exclamation_marks(x))

Признак 3. Доля причастий и деепричастий среди русских слов. Чем их больше, тем более книжным кажется текст

In [15]:
!pip install pymorphy2[fast]
import pymorphy2

     |████████████████████████████████| 51kB 4.2MB/s 
     |████████████████████████████████| 7.1MB 15.0MB/s 
     |████████████████████████████████| 256kB 55.4MB/s 
  Created wheel for DAWG: filename=DAWG-0.7.8-cp36-cp36m-linux_x86_64.whl size=767748 sha256=635538e5f0c6999f2305f359dc7296485551ba23a60cdfe896c3ba35c738efc6
  Stored in directory: /root/.cache/pip/wheels/d4/88/d0/4e4abc83eb8f59a71e8dbd8ba99fd5615a3af1fac1ef7f8825
Successfully built DAWG


In [0]:
morph = pymorphy2.MorphAnalyzer()
def n_participles(text, pymotphology):
    n = 0
    tokens = re.findall('[а-ёА-ЯЁ-]+', text)
    if tokens:
        n = 0
        for token in tokens:
            p = morph.parse(token)[0]
            pos = p.tag.POS
            if pos in ['PRTF', 'PRTS', 'GRND']:
                  n += 1
        p = n/len(tokens)
        return p
    else:
        return 0
     

In [0]:
data['n_participles'] = data['description'].apply(lambda x: n_participles(x, morph))

Признак 4. Наличие в тексте списков, написанных в столбик.
Предполагаем, что тексты отзывов со списками в столбик более структурированные и лучше воспринимаются. Искать будем разрывы строк после запятых и точек запятой либо числа/дeфисы в начале новой строки. Чтобы считать это списком, нужно хотя бы два однотипных указания на его наличие (списки из одного пункта не очень хорошие)

In [0]:
def contains_list(text):
    if len(re.findall('([,;])(\n)', text)) > 1:
        return 1
    elif len(re.findall('(\n)([0-9-])', text)) > 1:
        return 1
    else:
        return 0

In [0]:
data['list_in_text'] = data['description'].apply(lambda x: contains_list(x))

Признак 5. Наличие адреса в тексте. Предполагаем, что люди предпочитают договариваться о покупке, зная, куда конкретно за ней надо будет ехать

In [0]:
def adress_mentioned(text):
     adress_mentioned = 0
     if re.search('yл.?|улица|пр-т|проспект|бульвар|б-р', text):
         address_mentioned = 1
     return adress_mentioned

In [0]:
data['adress_in_text'] = data['description'].apply(lambda x: adress_mentioned(x))

Признак 6. Наличие фразы "в любое время/в любой день" (звонить/забирать/...) - такие

предложения подходят большему количеству покупателей

In [0]:
def any_time(text):
    if re.search('в любое время|в любой день', text):
        return 1
    else:
        return 0

In [0]:
data['any_time'] = data['description'].apply(lambda x: any_time(x))

In [0]:
from sklearn.model_selection import train_test_split

In [0]:
X = data[['text_len', 'text_len_tokens', 'number_of_caps', 'number_of_nonalphanum', 'size_info', 'question_and_exclamation', 'n_participles', 'list_in_text', 'adress_in_text', 'any_time']].values
y = data.deal_probability.values

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [0]:
from sklearn.linear_model import Ridge

In [45]:
regressor = Ridge(alpha=1)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

0.26320310910710787

От добавления метрик результат чуть-чуть ухудшился (с 4-й цифры после запятой).
Добавим векторайзер

In [0]:
from scipy.sparse import hstack

In [0]:
cv = CountVectorizer(max_features=1000, min_df=5, max_df=0.4, ngram_range=(1,2))
tfidf = TfidfVectorizer(max_features=1000, min_df=5,analyzer='char', ngram_range=(3,3), max_df=0.4)

X_cv = cv.fit_transform(data.description)
X_tfidf = tfidf.fit_transform(data.description)

X_with_CV = hstack([X_cv])
X_with_tfidf = hstack([X_tfidf])

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X_with_CV, y, random_state=42)

In [0]:
from sklearn.linear_model import SGDRegressor

In [53]:
regressor = SGDRegressor(max_iter=50, alpha=0.00001)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

0.25251747720006607

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X_with_tfidf, y, random_state=42)

In [55]:
regressor = SGDRegressor(max_iter=50, alpha=0.00001)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

0.2539645815677768

Пока что чуть-чуть лучше получается, если использовать tfidf. Соединим его с ранее использованными

In [0]:
Xtfidfcombo = hstack([X, X_with_tfidf])

In [0]:
X_train, X_test, y_train, y_test = train_test_split(Xtfidfcombo, y, random_state=42)

In [61]:
regressor = SGDRegressor(max_iter=50, alpha=0.00001)
regressor.fit(X_train, y_train)
preds = regressor.predict(X_test)
np.sqrt(mean_squared_error(y_test, preds))

48251853569864.23